cpp/src/io/parquet/page_enc.cu

```diff
@@ -1337,11 +1343,206 @@ static __device__ void byte_reverse128(__int128_t v, void* dst)
+ */
+static __device__ uint32_t truncate_binary(device_span<uint8_t const> arr,
+                                           const void** res,
+                                           bool is_min,
+                                           void* scratch,
+                                           size_type truncate_length)
+{
+  if (is_min) {
+    *res = arr.data();
+    return truncate_length;
+  }
+  memcpy(scratch, arr.data(), truncate_length);
+  // increment last byte, working backwards if the byte overflows
+  auto const ptr = static_cast<uint8_t*>(scratch);
+  for (int32_t i = truncate_length - 1; i >= 0; i--) {
+    ptr[i]++;
+    if (ptr[i] != 0) {  // no overflow
+      *res = scratch;
+      return i + 1;
+    }
+  }
+
+  // couldn't truncate, return original value
+  *res = arr.data();
+  return arr.size_bytes();
+}
+
+// TODO (ets): the assumption here is that string columns might have UTF-8 or plain binary,
+// while binary columns are assumed to be binary and will be treated as such.  If this assumption
+// is incorrect, then truncate_byte_array() and truncate_string() should just be combined into
+// a single function.
+/**
+ * @brief Truncate a UTF-8 string to at most truncate_length bytes.
+ */
+static __device__ uint32_t truncate_string(
+  const string_view& str, const void** res, bool is_min, void* scratch, size_type truncate_length)
+{
+  if (truncate_length == NO_TRUNC_STATS or str.size_bytes() <= truncate_length) {
+    *res = str.data();
+    return str.size_bytes();
+  }
+
+  // convert char to uint8_t since UTF-8 is just bytes, not chars.  can't use std::byte because
+  // that can't be incremented.
+  auto const span =
+    device_span<uint8_t const>(reinterpret_cast<uint8_t const*>(str.data()), str.size_bytes());
+
+  // if str is all 8-bit chars, or is actually not UTF-8, then we can just use truncate_binary()
+  if (str.size_bytes() != str.length() and is_valid_utf8_string(span.first(truncate_length))) {
+    return truncate_utf8(span, res, is_min, scratch, truncate_length);
+  }
+  return truncate_binary(span, res, is_min, scratch, truncate_length);
+}
+
+/**
+ * @brief Truncate a binary array to at most truncate_length bytes.
+ */
+static __device__ uint32_t truncate_byte_array(const statistics::byte_array_view& arr,
+                                               const void** res,
+                                               bool is_min,
+                                               void* scratch,
+                                               size_type truncate_length)
+{
+  if (truncate_length == NO_TRUNC_STATS or arr.size_bytes() <= truncate_length) {
+    *res = arr.data();
+    return arr.size_bytes();
+  }
+
+  // TODO: this will need to change when #11424 merges
+  device_span<uint8_t const> const span{arr.data(), arr.size_bytes()};
+  return truncate_binary(span, res, is_min, scratch, truncate_length);
+}
+
+static __device__ void get_extremum(const statistics_val* stats_val,
```

Comment:
Docs please.

Reply:
In addition, does this function find just one value for the output? If so, please return the value, don't modify the input pointer.

Reply:
changed to return std::pair.  Thanks, it's much cleaner now!